addons/web/static/src/core/template_inheritance.js

```diff
@@ -0,0 +1,244 @@
+    if (node.tagName === "xpath") {
+        const xpath = getXpath(node);
+        const result = doc.evaluate(xpath, root, null, XPathResult.FIRST_ORDERED_NODE_TYPE);
+        return result.singleNodeValue;
+    }
+    for (const elem of root.querySelectorAll(node.tagName)) {
+        if (
+            [...node.attributes].every(
+                ({ name, value }) => name === "position" || elem.getAttribute(name) === value
+            )
+        ) {
+            return elem;
+        }
+    }
+    return null;
+}
+
+function getNodes(target, node) {
+    const nodes = [];
+    for (const childNode of node.childNodes) {
+        if (childNode.tagName === "xpath" && childNode.getAttribute?.("position") === "move") {
+            const node = getNode(target, childNode);
+            removeNode(node);
+            nodes.push(node);
+        } else {
+            nodes.push(childNode);
+        }
+    }
+    return nodes;
+}
+
+function mergeTextNodes(first, second, trimEnd = true) {
+    first.data = (trimEnd ? first.data.trimEnd() : first.data) + second.data;
+    second.remove();
+}
+
+function removeNode(node) {
+    const { nextSibling, previousSibling } = node;
+    node.remove();
+    if (nextSibling?.nodeType === Node.TEXT_NODE && previousSibling?.nodeType === Node.TEXT_NODE) {
+        mergeTextNodes(
+            previousSibling,
+            nextSibling,
+            previousSibling.parentElement.firstChild === previousSibling
+        );
+    }
+}
+export function applyInheritance(source, operations, url = "") {
+    for (const operation of operations.children) {
+        const target = getNode(source, operation);
+        if (!target) {
+            throw new Error(`Element '${operation.outerHTML}' cannot be located in parent view`);
+        }
+        const position = operation.getAttribute("position") || "inside";
+
+        if (odoo.debug && url) {
+            const attributes = [...operation.attributes].map(
+                ({ name, value }) =>
+                    `${name}=${JSON.stringify(name === "position" ? position : value)}`
+            );
+            const comment = document.createComment(
+                ` From file: ${url} ; ${attributes.join(" ; ")} `
+            );
+            if (position === "attributes") {
+                target.before(comment); // what about modification of root?
+            } else {
+                operation.prepend(comment);
+            }
+        }
+
+        switch (position) {
+            case "replace": {
+                const mode = operation.getAttribute("mode") || "outer";
+                switch (mode) {
+                    case "outer": {
+                        const result = operation.ownerDocument.evaluate(
+                            ".//*[text()='$0']",
+                            operation,
+                            null,
+                            XPathResult.ORDERED_NODE_ITERATOR_TYPE
```

Comment:
I think you can get rid of some complexity if you use [ORDERED_NODE_SNAPSHOT_TYPE](https://developer.mozilla.org/en-US/docs/Web/API/Document/evaluate#ordered_node_snapshot_type), you won't need to collect the nodes by hand from the iterator.

Reply:
Seems to work well